### ISSUE TITLE
Converge External State (EBS attachments, elastic IP, load balancer, etc) externally

### ISSUE THREAD

**Issue by mrflip** [infochimps-labs/ironfan#23]
says @howech,

Folks,

I would like to propose a change to the way ebs volumes get mounted.

In the current scheme, there is an ebs volume mounter chef script that runs as a part of the chef-client run. It pulls data out of the cluster_ebs_volumes data bag, and does an excellent good job of mounting volumes in the right places.

But if it works so well, why change it?

Security. 

Because of this recipe, we have to put a too-general AWS key on all of our EC2 instances. Should a node really be given the power to manage its own AWS EC2 destiny - maybe.  Should it also be given the ability to manage the destiny of other nodes - probably not. However, the idea of setting up custom api-keys, security groups and policies that make it so that the key on bonobo-slave-17 is only given power over bonobo or the bonobo-slave facet, or even the bonobo-slave-17 node seems extraordinarily complicated.

Instead, imagine a world where "knife cluster launch bonobo" would look into the same databag and pull out the mapping of instance names to ebs volumes. (Ok, bad example, since bonobo does not have ebs volumes, but you get the idea).  Most of the time, you just want to make sure that the ebs volumes get set up in the right place when the node is started: I _think_ that an ebs volume will remain attached to an instance through anything you can do to the instance (except of course to detach it, and to terminate the node...). Mounting the device does not require AWS EC2 privs, so that part of the process can remain the way it is. If we think we need it, we could maybe write something to make sure that a running cluster has its ebs volumes attached in the right places, and optionally fix it.  The "fix it" part is problematic, as you have to ensure that the drives are unmounted if you want to detach them - as this could fatally interfere with running processes on a server, it is probably best to keep this part manual.  The rest should be relatively simple to do with Fog.

Comments?

**Comment by mrflip** [comment]
There's a possibility -- I don't know, just something I noticed -- that you can specify volumes to attach when you _launch_ an instance. Worth looking at, and possibly worth adding to Fog if it's missing.

(paging @temujin9 @aseever)

**Comment by mrflip** [comment]
So the philosophical debate for me is

> "Is the mutable configuration of a machine -- its IP address, EBS volumes, ELB adherence, AWS console tags, etc -- something that chef should control, or that an external process should control?"

In practical terms, the goals of a good solution:
- If I launch a cluster and things go well, no additional action is needed: a reliable, predictable process will converge the machine's external and internal configuration to its declared goal state.
- idempotent-from-coherent state: I can at any time invoke action X and, unless things are fubar, it will converge or fail. Currently action X is `sudo chef-client`. X could also be running `knife cluster converge`, or something else.
- If things are screwed up (a volume is attached to the wrong instance, or an instance isn't launched) the things that are not screwed up are given a chance to succeed, and that it leaves me in no doubt that my world is fucked up, how and why.

## \* There is a clear division of responsibility btwn chef and the machine, or reason to believe that idempotency means I don't have to sweat it.
- given all of the above, I'd love to live in a world where a person with root access did not therefore have access to a powerful AWS key.

One way to think about it would be

> "Hey, we have a thing (called "chef") that is _really good_  at examining a system and converging it to a declared final state. Let's go ahead and use that thing. The benefits:
> - a machine always controls its destiny 
> - after the initial DNA is transferred, all knowledge about the final state lives in chef
> - when chef is run, the world is right -- there's no outside influence that could fuck up the initial or final or in-transit state.

The other way to think about it is

> "Chef is for the _internal_ configuration of a machine, and trying to use it to  manage the external configuration is revealing impedance mismatches.
> - if a machine is fubar, we can't depend on it fixing itself
> - the aforementioned security issues
> - others, I'm sure.
>   Meanwhile, the convergence logic we're talking about is small, and unexpected to grow:
> 
> ```
>   HAI
>   HAZ TEH ELASTIC_IP my_elastic_ip?, O RLY?
>   YA RLY: VISIBLE "HAZ ELASTIC_IP"
>   NO WAI: O HAI "AWS" ELASTIC_IP_GIMMEH my_elastic_ip
>   OIC
> ```
> 
> (disappointed to find there is no lolcode syntax highlighter) 
> knife is the tool we're using to manage a cluster's state, so that's the tool we should use.

Several of the things we've listed depend on the machine reaching a certain state in its lifecycle before they're appropriate:
- IP address: either started or starting, don't remember
- EBS volumes: machine must be started (?)
- ELB adherence: mostly harmless to have it added anytime, but there's some appeal to having it attach only when the relevant service goes green.
- AWS console tags: can happen anytime AFAIK
  Lifecycle dependencies argue for the machine to trigger them when appropriate -- this makes things more dependable in the general case (yay chef), more brittle in the worst case (yay knife).

Finally, some of these adhere to the machine (elastic IP, EBS volumes, tags), others to the recipe (ELB, tags), which argues for chef.

All of which is to say... sounds like a tie. But if both knife and chef are idempotent and reliable when their world is right, and both burst into flames when fubar, I like the idea of "knife cluster converge".

**Comment by howech** [comment]
Here is a question for you:

When would you ever REALLY want chef-client to change the ebs volumes
attached a live server?

The scenario here is that you have a cluster running, and you change it's
ebs volumes in the static cluster definition (in this case, stored in
cluster_ebs_volumes databag). For chef-client to get the server to converge
to its definition, it has to do the following steps:
1. stop any processes that have open files on the mounted volume (because
   unmounting a directory with open files is unfriendly)
2. unmount the volume (because detaching a mounted ebs volume is
   impossible)
3. detach the existing ebs volume
4. (if you are reshuffling multiple volumes among servers, you have to
   wait until the new volume is available)
5. attach the new ebs volume
6. make sure that the drive is formatted the way you expect it to be
7. mount the volume
8. restart the processes from step 1

There are lots of plates spinning here, of course. I guess you could assume
the that authors of the cluster configuration (cluster_ebs_volumes databag)
knew what they were doing, so step 6 may be unnecessary. Overall, though,
the remainder would be necessary for a "reattach" type of script, and it
introduces a number of problems - how do you go about managing the
dependencies between the various services and the mounted volume?  How do
you keep accidental changes to the cluster_ebs_volumes databag from stopping
production services?

I hope that the answer to the "when" question a couple of paragraphs ago is:
"NEVER!"  (i would of course accept "never on anything that I really care
about" as well...)

The complexity of the the 8 step road to disaster above is such that I would
want to make sure that a human was in control of the process.

( I guess when there is not a currently attached volume, much of the
complexity goes away, but does that every really happen outside of creating
a new cluster? Not often! )

I am happy with the idea of a "knife cluster converge" task, that would go
through the entire cluster definition and make sure that the reality
coincided with the definition, as long as it had an appropriate amount of
"do you really want to do this" interaction tacked on to it.

<straw_man>

On the other side of the argument, suppose that you really did want
chef-client to be able to administer the ec2 aspects of a server. To make
this adhere to the "least privilege" security principle, you would want the
credentials it possesses to only allow it what it needs to do.  This would
involve creating a new AWS IAM security credential for every server that
only has access to modifications to itself (or perhaps its facet or
cluster).

I think that this could be possible, as I think I saw some IAM support in
Fog, but I suspect that it will be rather complicated.

</straw_man>

Basically, I am in favor of the internal/external configuration approach,
mostly because it simplifies the security picture. You can't entirely
eliminate the external tool for bootstrapping: that is why we wrote
cluster_chef to begin with. Since it has to be there, you may as well
leverage it. "knife cluster converge" seems like a perfect solution.

--Chris

**Comment by mrflip** [comment]
+1 on implementing knife cluster converge

On Mon, Jun 6, 2011 at 4:34 PM, howech <
reply@reply.github.com>wrote:

> Here is a question for you:
> 
> When would you ever REALLY want chef-client to change the ebs volumes
> attached a live server?
> 
> The scenario here is that you have a cluster running, and you change it's
> ebs volumes in the static cluster definition (in this case, stored in
> cluster_ebs_volumes databag). For chef-client to get the server to converge
> to its definition, it has to do the following steps:
> 1. stop any processes that have open files on the mounted volume (because
>    unmounting a directory with open files is unfriendly)
> 2. unmount the volume (because detaching a mounted ebs volume is
>    impossible)
> 3. detach the existing ebs volume
> 4. (if you are reshuffling multiple volumes among servers, you have to
>    wait until the new volume is available)
> 5. attach the new ebs volume
> 6. make sure that the drive is formatted the way you expect it to be
> 7. mount the volume
> 8. restart the processes from step 1
> 
> There are lots of plates spinning here, of course. I guess you could assume
> the that authors of the cluster configuration (cluster_ebs_volumes databag)
> knew what they were doing, so step 6 may be unnecessary. Overall, though,
> the remainder would be necessary for a "reattach" type of script, and it
> introduces a number of problems - how do you go about managing the
> dependencies between the various services and the mounted volume?  How do
> you keep accidental changes to the cluster_ebs_volumes databag from
> stopping
> production services?
> 
> I hope that the answer to the "when" question a couple of paragraphs ago
> is:
> "NEVER!"  (i would of course accept "never on anything that I really care
> about" as well...)
> 
> The complexity of the the 8 step road to disaster above is such that I
> would
> want to make sure that a human was in control of the process.
> 
> ( I guess when there is not a currently attached volume, much of the
> complexity goes away, but does that every really happen outside of creating
> a new cluster? Not often! )
> 
> I am happy with the idea of a "knife cluster converge" task, that would go
> through the entire cluster definition and make sure that the reality
> coincided with the definition, as long as it had an appropriate amount of
> "do you really want to do this" interaction tacked on to it.
> 
> <straw_man>
> 
> On the other side of the argument, suppose that you really did want
> chef-client to be able to administer the ec2 aspects of a server. To make
> this adhere to the "least privilege" security principle, you would want the
> credentials it possesses to only allow it what it needs to do.  This would
> involve creating a new AWS IAM security credential for every server that
> only has access to modifications to itself (or perhaps its facet or
> cluster).
> 
> I think that this could be possible, as I think I saw some IAM support in
> Fog, but I suspect that it will be rather complicated.
> 
> </straw_man>
> 
> Basically, I am in favor of the internal/external configuration approach,
> mostly because it simplifies the security picture. You can't entirely
> eliminate the external tool for bootstrapping: that is why we wrote
> cluster_chef to begin with. Since it has to be there, you may as well
> leverage it. "knife cluster converge" seems like a perfect solution.
> 
> --Chris
> 
> ## 
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/infochimps/cluster_chef/issues/23#comment_1312064

## 

infochimps.com - discover data

**Comment by temujin9** [comment]
+1 for clear separation of concerns (internal -> chef, external -> knife cluster).

**Comment by howech** [comment]
I added the ability to specify external volumes within the cluster specification.  These will be attached as a part of cluster launch. 

cluster "x" do
  facet "y" do
    server "0" do
      volume :id => "vol-12312312", :device => "/dev/sdk"
    end
  end
end

This will attach volume vol-123123123 on device /dev/sdk on the machine whose default chef node name is "x-y-0".

**Comment by mrflip** [comment]
fixed by `knife cluster sync`
